Add functionality for getting atomic projections for correct spectral function plotting

kavanase opened the pull request:
I've now added functions and CLI options to (1) obtain the atomic projections for the electronic states, and (2) then plot these as an RGB colour map on the spectral function.

The atomic projections are obtained by simply specifying `--natoms` when calling `easyunfold unfold calculate`, and optionally also `--procar_path` if your `PROCAR` has been renamed or is in a different directory:
![image](https://user-images.githubusercontent.com/51478689/161347624-238ff0de-c0b3-4a8d-9229-09e1f2001649.png)
I've added `tqdm` timing of this, but it is typically far quicker than the initial `WAVECAR` unfolding. This saves the atomic weight info to the `json` file.

If the `--atomic_proj` option is then included when calling the `plot` command, the spectral function will then be plotted as an RGBA colour map, with the atomic projections controlling the colour (RGB) and the spectral density controlling A (alpha; opacity). Using `easyunfold unfold plot --eref 4.196 --emin -4 --emax 4 --atomic_proj` with NaBiS2 for example, the result is this:
![projected_spectral_unfold](https://user-images.githubusercontent.com/51478689/161348009-b45f15f7-900c-4f6f-9224-eda7d0f3b689.png)

For comparison to the corresponding DOS: (apologies for the confusion of using green for Bi above but for S below, and vice versa with blue 😅)
<img width="288" alt="image" src="https://user-images.githubusercontent.com/51478689/161349916-b6a4cb73-3b81-4e8b-9995-0ff14afa52fc.png">

Original un-projected: (`BuPu` `cmap`)
![cmap_BuPu_unfold](https://user-images.githubusercontent.com/51478689/161348071-d514bb3a-8d13-47a2-9149-21860a03ff5a.png)

Currently, this restricts the user to using atomic projections for only 2 or 3 species. If there are only two atomic species, the green channel is set to zero and the colour map is from red to blue: 
![fake2atom_unfold](https://user-images.githubusercontent.com/51478689/161348285-72fa4694-394a-475d-8726-afb8a044255b.png)
(Example here generated by creating a modified `json` with Na removed from the atomic projections array) Looks pretty fkn sick if you ask me!
Only being able to plot the projections for 3 species at a time I think is absolutely fine. This is the case in `sumo` also, and there is very rarely a situation where you really want to show more than 3 species projections on a single band structure. In those rare cases, you're probably better off doing a side-by-side band structure & DOS anyway.

Tested and all seems to work fine and as expected.

So useful additions I think could be:
- Pulling the number of atoms and identities from the vasprun.xml, so that the user just needs to specify which 2 or 3 species they want to project (as in `sumo`), and then done automatically from there. Would also then add a legend for the colour projections to the plot.
- Allowing for orbital (rather than just atomic) projections? Should be relatively straightforward to extend to this, and can template off / use pymatgen code for this
- Add an option to also plot a DOS alongside the bandstructure, as in and using the functions from `sumo`.

zhubonan commented:
Sorry somehow this PR has slipped through my to-do list. 

I remember implementing plotting multiple graphs using different weights but with the same color map. So one would get a row of plots instead of a single plot with differnet colour maps. This would allow any number of species to be supported.

But I think having a single point is also neat. Will see if I can add this in. The code has diverged in the dev branch so I won't be able to merge it.

kavanase commented:
Sounds good @zhubonan! 
Let me know if this gets implemented, and I can knock out an example workflow & results matching the above, to stick on the docs and add tests for it as well

zhubonan commented:
@kavanase I have just added #8 with implementation to allow custom terminal colours as in `sumo-bandplot`.  Sorry that is has been a (long) while since I last worked on it .

In #8, the algorithm for calculating the intensity and colouring of the spectral function is slightly different compared to here. There, the spectral functions are calculated using atomic-projection-weighted weights directly, and then the final spectral function is coloured by weighting the spectral function of each selected channel. 

Compared to the method in this PR, where each pixel is coloured by searching for the projected weights of the nearest band, it has the advantage for allowing only a subset of the projections to be used (which can affects the intensity of the spectral function itself) and it also takes account of projections from multiple supercell kpoints at a single primitive cell kpoint (due to reduced symmetry).

Please take a look of #8. Once we have it merged this PR can be closed I think.

kavanase commented:
Here I've rebased onto `dev` and then added the last commit in that list. This was to change the `--help` behaviour from:

![image](https://user-images.githubusercontent.com/51478689/219871018-5606ae34-ecc0-4066-8639-1806808aa18c.png)

Where `atoms-idx`, `cmap` and `procar` are duplicated, and also the default behaviour for some parameters are not shown (but would be useful for the user to know, to then decide what and how to change to customise the plot). This is now:

![image](https://user-images.githubusercontent.com/51478689/219871103-11bd6770-a2c4-4782-9c83-3589d0509fe7.png)

Likewise the help message for `easyunfold unfold plot --help` is thus also updated to show the useful default values:
Original:
![image](https://user-images.githubusercontent.com/51478689/219871069-a8734f14-b176-4971-a7c5-5d341b937eb9.png)
Now:
![image](https://user-images.githubusercontent.com/51478689/219871168-baa21917-5975-47e0-935c-5c1bf3c2c5b3.png)

kavanase commented:
[NaBiS2 SQS Projection - easyunfold Docs.pdf](https://github.com/SMTG-UCL/easyunfold/files/10774530/NaBiS2.SQS.Projection.-.easyunfold.Docs.pdf)

Just attaching the printed new docs page here so you can see (roughly) what it looks like

zhubonan commented:
Docs looks great! Let me know when it is ready to merge 🚀

zhubonan commented on docs/example_nabis2.md:
```diff
@@ -0,0 +1,89 @@
+
+Note that the path of the `PROCAR` is passed along with the group of atoms.
+In this example, the first 20 atoms are `Na`, the second 20 are `Bi` and the last 40 are `S`. Different groups are 
+separated by `|`, and `-` can be used to define the range.
```
I thought `,` is the separator here? We can support both with `re.split(r'[,|]', string)`.

kavanase commented:
So these updates work locally, with these outputs, but may require some tuning of the GH Actions workflow. Let's see...

[easyunfold.plotting - easyunfold Docs.pdf](https://github.com/SMTG-UCL/easyunfold/files/10777130/easyunfold.plotting.-.easyunfold.Docs.pdf)

[CLI Commands - easyunfold Docs.pdf](https://github.com/SMTG-UCL/easyunfold/files/10777131/CLI.Commands.-.easyunfold.Docs.pdf)

<img width="1512" alt="Screenshot 2023-02-19 at 14 29 31" src="https://user-images.githubusercontent.com/51478689/219954737-6535fbc6-eb46-4a47-bf75-95e9135fc12d.png">

<img width="1512" alt="Screenshot 2023-02-19 at 14 29 16" src="https://user-images.githubusercontent.com/51478689/219954727-034a8530-bec8-4610-b584-0b2ac6654d44.png">

kavanase commented:
Yep! Does the job: https://smtg-ucl.github.io/easyunfold/cli/ 👌

kavanase commented:
This passes all tests (bar a "0.12% decrease in code coverage") and docs look good, happy for me to merge @zhubonan?

zhubonan reviewed (approved):
All look good! Thanks 🚀

zhubonan commented:
> This passes all tests (bar a "0.12% decrease in code coverage") 

This is fine. For the documentation, I guess at the moment one has to manually regenerate the API docs with python automate_mkdocs.py`? I think this is OK.

zhubonan commented:
Merged 👏 👏

kavanase commented:
Just replying to this here:
> For the documentation, I guess at the moment one has to manually regenerate the API docs with `python automate_mkdocs.py`? I think this is OK.

Well sort of. So `python automate_mkdocs.py` populates the `mkgendocs.yml` file which specifies which functions to generate API docs for, and then `gendocs --config mkgendocs.yml` generates these API docs. So if the docstrings of one of these functions are changed, then only `gendocs --config mkgendocs.yml` needs to be run to update with this. But if a new function is added then API docs won't be generated for it until `python automate_mkdocs.py` is run or it's manually added to `mkgendocs.yml`, before running `gendocs --config mkgendocs.yml`.

I got this workflow from https://towardsdatascience.com/easily-automate-and-never-touch-your-documentation-again-a98c91ce1b95. Should be easy to set it up as GH Action though, with just these two commands (and checking it doesn't mess with any of the other docs).
